spec/lita/handlers/github_prs/chat_handler_spec.rb

```diff
@@ -74,4 +74,54 @@ def send_attachments(destination, attachments)
     let_context(body: "  \n \t  foo bar \n \n baz \t ") { it { should eq [] } }
     let_context(body: nil) { it { should eq [] } }
   end
+
+  describe "#additional_todos" do
+    it "generates a list of todos in markdown via a generator" do
+      module Organization
+        class ShortName
+
+          def initialize(_params); end
+
+          def additional_todos_markdown
+            "this is markdown"
+          end
+        end
+      end
+      chat_handler = Lita::Handlers::GithubPrs::ChatHandler.new(robot)
```

Comment:
I'm not sure I agree. To me implementation details are often relevant whenever I need to understand what a test is meant to test. 

In this example we could save some characters, sure. It will however come with some costs:

1. Readers will need to scroll up and down to know what is being initialized and how
2. Risk that a change will affect tests they aren't supposed to change
3. Relying on a macro that is basically a function wrapper (which means more complexity for developers without rspec experience)

Following the DRY principle in application code makes sense at it makes for easier maintainability but I don't think following it in the same way makes sense for tests.

I'd love to share more thoughts on the matter if you'd like!

I can make use of `let` for the sake of consistency if you'd like?

Reply:
Yes you need to scroll up to see it, but to me the question should be how often do you need to scroll up?  Thinking about this that part is what's a bit clearer to me about what makes this eligible for moving up.  I've been thinking about [this talk](https://www.youtube.com/watch?v=URSWYvyc42M) by Sandi Metz recently and she explained about mocks (stick with me a minute ;) ) that what you're doing with a mock is making a bet that the API of the object outside the code under test isn't going to change, at least not as much as the code that otherwise would be executing if you hadn't mocked.  In this case, I would argue that this line is not likely to change much and that most of the time you won't need to refer to it because initializing the chat handler is an implementation detail.

Also, to me it's not about saving characters but making the code easier to read at a glance.  Maybe that's the same thing in your mind, but I just wanted to clarify that I don't care so much about copy/pasting as I do about the cost of needing to read through things to parse a test.

There's also the benefit of DRYing code which I people don't talk about very often: When code is only expressed one time, that is an assertion that the writer of the code is making saying that they not doing something different every time.  If you copy/paste that means that the reader of the code needs to re-evaluate that code each time they look at it and verify themselves that it's doing the exact same thing as the previous code which looked very similar.  It also leaves room for drift of each version of the copied code.  By putting the code in one place it forces you to express in your test suite the times when your setup is different, if it's different at all.

Regarding number 2. yes, it could certainly affect other tests.  That's an argument for establishing it from the beginning, but it's a risk that you need to weigh which refactoring.

And regarding number 3. sure it can involve complexity, but so can everything that we write.  It's our responsibility to use the tools we have in what we think is a responsible way.  Just like we should not be afraid of using big words with our children, we shouldn't be afraid of using tools which others aren't familiar with.  It provides an opportunity for growth, especially if we can be open to helping them.

I don't want you to use `let` until I've convinced you ;)

Reply:
I actually find myself having to scroll up and down more often than not, especially when there are contexts where variables sometimes get overwritten and sometimes not. Sometimes one `let` is dependent on multiple others which makes the path to finding the actual structures a test makes use of. 
Add `subject` on that and the context of simple assertions could potentially be spread out on multiple hundred lines depending on the size of the testfile. 

I'm generally open to extracting parts of a test whenever there are clear gains to do so. I however tend to ignore the rspec parts and instead make very descriptive helpers. Factories are also helpful but i

```ruby
it "requires some heavy setup" do
  set_configuration(show_errors: true)

  expect(fake_class)
end 

def set_configuration(**config)
  config.each{ |key, value| allow(Rails.application.config).to receive(key).and_return(value) }
end
```


I'm sorry but I don't think convincing me about that is possible as I was there but was convinced the other way around :(

Reply:
I'm not disagreeing that using a lot of `let`s can make for a complicated test file.  But I think that there is still room in between.  All I'm saying for now is that this one particular `let` seems like an ideal candidate for moving.  If you think that you would need to keep scrolling up to see the `chat_handler` definition, then that's one thing, but I suspect you don't ;)